src/Compiler/Scheme/ChezSep.idr

```diff
@@ -0,0 +1,338 @@
+        ++ "\"$@\""
+    ]
+
+startChezCmd : String -> String -> String -> String
+startChezCmd chez appDirSh targetSh = unlines
+    [ "@echo off"
+    , "set APPDIR=%~dp0"
+    , "set PATH=%APPDIR%\\" ++ appDirSh ++ ";%PATH%"
+    , "\"" ++ chez ++ "\" -q "
+        ++ "--libdirs \"%APPDIR%/" ++ appDirSh ++ "\" "
+        ++ "--program \"%APPDIR%/" ++ targetSh ++ "\" "
+        ++ "%*"
+    ]
+
+startChezWinSh : String -> String -> String -> String
+startChezWinSh chez appDirSh targetSh = unlines
+    [ "#!/bin/sh"
+    , ""
+    , "set -e # exit on any error"
+    , ""
+    , "DIR=$(dirname \"$(realpath \"$0\")\")"
+    , "CHEZ=$(cygpath \"" ++ chez ++"\")"
+    , "export PATH=\"$DIR/" ++ appDirSh ++ "\":$PATH"
+    , "\"$CHEZ\" --program \"$DIR/" ++ targetSh ++ "\" \"$@\""
+    , "\"$CHEZ\" -q "
+        ++ "--libdirs \"$DIR/" ++ appDirSh ++ "\" "
+        ++ "--program \"$DIR/" ++ targetSh ++ "\" "
+        ++ "\"$@\""
+    ]
+
+-- TODO: parallelise this
+compileChezLibraries : (chez : String) -> (libDir : String) -> (ssFiles : List String) -> Core ()
+compileChezLibraries chez libDir ssFiles = ignore $ coreLift $ system $ unwords
+  [ "echo"
+  , unwords
+    [ "'(parameterize ([optimize-level 3] [compile-file-message #f]) (compile-library " ++ chezString ssFile ++ "))'"
+    | ssFile <- ssFiles
+    ]
+  , "|", chez, "-q", "--libdirs", libDir
+  ]
+
+compileChezLibrary : (chez : String) -> (libDir : String) -> (ssFile : String) -> Core ()
+compileChezLibrary chez libDir ssFile = ignore $ coreLift $ system $ unwords
+  [ "echo"
+  , "'(parameterize ([optimize-level 3] [compile-file-message #f]) (compile-library " ++ chezString ssFile ++ "))'"
+  , "|", chez, "-q", "--libdirs", libDir
+  ]
+
+compileChezProgram : (chez : String) -> (libDir : String) -> (ssFile : String) -> Core ()
+compileChezProgram chez libDir ssFile = ignore $ coreLift $ system $ unwords
+  [ "echo"
+  , "'(parameterize ([optimize-level 3] [compile-file-message #f]) (compile-program " ++ chezString ssFile ++ "))'"
+  , "|", chez, "-q", "--libdirs", libDir
+  ]
+
+writeFileCore : (fname : String) -> (content : String) -> Core ()
+writeFileCore fname content =
+  coreLift (writeFile fname content) >>= \case
+    Right () => pure ()
+    Left err => throw $ FileErr fname err
+
+readFileCore : (fname : String) -> Core String
+readFileCore fname =
+  coreLift (readFile fname) >>= \case
+    Right content => pure content
+    Left err => throw $ FileErr fname err
+
+chezNS : Namespace -> String
+chezNS ns = case showNSWithSep "-" ns of
+  "" => "unqualified"
+  nss => nss
+
+-- arbitrarily name the compilation unit
+-- after the alphabetically first namespace contained within
+chezLibraryName : CompilationUnit def -> String
+chezLibraryName cu =
+  case SortedSet.toList cu.namespaces of
+    [] => "unknown"  -- this will never happen because the Tarjan algorithm won't produce an empty SCC
+    ns::_ => chezNS ns
+
+touch : String -> Core ()
```

Comment:
Same: can we move this to `Core.Core`? Maybe even create `Core.Core.File` if we have a lot of file manipulation functions.